### PR TITLE
Don't put a space before a prefix operator, unless we have to.

### LINKED ISSUE #3131: Unary operators: inconsistent formatting when arg is literal vs variable
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/main?data=N4KABGBEDGD2AmBTSAuKAbRAXMAPMAvGAH7ECMAOgE4UB2mOAnoScbpADThQDO2WAS1oBzHqjABtbhFAQ5UACRZGAB2RpIQrJ2lzICgG4BDdAFd1k3fIAMXefMgBJWklpYAygIBeyO-YgALFZgALq6AL5%2BMsH6ymrimm46-vrGZhZS-hBkUfaQALJGuAAyQojFiCJYABbJWWQATNbBYfKRurJ5SqoWkJXwAPIAZqW0iO7KmHUOhibm4plZDbkOAKIuw6O%2BwRCQ6EOQLRG5nTNxvQBGsLDo03qz6Qs7YADMK3rOfFRYAGJCJgA5RAAd3QZTu8iwVHMRzaJxi3XiGiuNwhijS8zQi38AXeu3cKiM0EQACFEENYFREAAFIxUIwAW2wiCoaIgUJh-lacna8lO93OCRRt3eqTmGWeAFY8VACUTSeTKeVYMCWdAjHxnAZYOrBLBaGywBzELCefCUojLtcRQiMRKsmAAGwyyBy4lkilUgCqKjUVHVmto2t1An1hqGJj4pogvLk-N2lqF1rRYseWOeAHYXW6FZ7EABhdAanj5-U8DnQLCU8ORk1c44dBGC5HJ0UPTGWB0ADmzhPdiqp%2BUQDIuLJr6Cj9bhjYtzagwpT7ftWQAnL35R6laX0GGZRGJ3X7NyY%2BaunPIAu23ang6yLZnq6%2B4gAIJDLAs0sMhlGQ3G6NgWNolnHok1RK9xRveocgfHNNypdxhwEOAdwNPda3-QCwHjRRz0vW0IPTW9lhgp9X3fKgEIZJCbl3Z4-ynM0ZzPECWzA-C007eo3hI%2BVnyoWBTBcAARRAwSo8jf2hQ95GPADTzOFioC0Rdr0I%2BpcQfQpcEcIYABVqkqdxqkpLAAHUBHgGpDWaBiTyYhSkSUpJwI47F7DIaVNKKHT9MqVYDyMkzzMs2oZUdGyjwbPkm0UxJtBcjs3PkMhnS87TaCGARcAGP0jCrKhVlwFQqR4HhQ1Q54uwimSorjGLHLilSCM4-wyCzNKACVEDgKh4GCqyZQCaq5FkzDsNiWLlIS5dWp7TruspeAAVMEcWWGRx3wZMQZTIDD5IFWKGVMdBBDBMYfkpb8sHI3SQOmyDWrXB8up6%2BB8mO06yguqgrokl1oGqOkiXIgB9YELKsvb7IOhqpvYxLniaF0tL4%2BlGAGKhSnLfrQsq4aIFG-aE3POGLVUlr7AaaCHQKIpUaMdHMYEctltWqh1s27bnl22y5Oh4nDo%2BgQzsQb7fpZW74nutT-AaYiafpxmsawd6TuFr7Lryv6HwBoHKxZMGIdqKHouA2HnPhmbKe4mmtIANXFEkhHgIRhBxw0qpNuqzd6UmunJpK5AaDTbaKH5BMrcqnZcV33cG-HQlqoDmPN%2BLLYeynPND3AhNgLAAHFsEK4rEFK8q47xr3k4cq02LJ5rA4gBpUppoFQTKMksFVSpJcQETMtoAQ9VoZ8XCHNmuYdejIunU2U9rm169cxH2oVsFhFocPaEj-VPA3vLTCpXTYGcVwsDy8rx0nGfGLnmvQMX-2G8Rua1-I2gL4MZURHH0cqH7oQQ9yqTyyPua%2BNVZ7e3ng-Jqy8HQNCerbIWWxiiMguPAIwhZYBlREG3EWV9pIjSTlheqC9YEIwdC8e8NNi4sgEEyNwJgADSiBEAqFPpULAzgST0m3rjB0YDCEE2IeNRMrFH4zADs8F41MsiQBJEWWgABrLYPA%2BICRcECcsiA3pCxFsXEqZUyySU5DfOyd8YZkOlhTeQLx5ZyJJHSOCfdmbQCoPQ-4754BekHvqES0Aiz0mHgQquJCfYJCOmrEWPCiRKOwBMRgUxrGN1eDbORqtPpjBidAOJHhJjbBpm4xkah4CHF5mNUhCQ-aSOfpQkOciWFsK0qzP%2BwwFFGGUaow0d4E6E35jhWKeEl4UKyC8LOci8EdwHIgDJ6sxifhUKYc%2Bw8DGlyMRVKeUlQmiNwq2dOMt7AvBbnI2h7iGHn3QP5KiPB%2BGgPQuU7gIQ-CaB4D8Mq4hBEgHCEAA)

(This is @Smaug123 but from the office.)

#### Code

```fsharp
let x = ~~1
let y = ~~x
```
            

#### Result

```fsharp
let x = ~~ 1
let y = ~~x

```
            
#### Problem description

The formatting is inconsistent between when the argument of a unary operator is a literal vs a variable. I'd prefer us to stick with `~~1` (removing the space, as we do with the variable case).

#### Extra information

- [ ] The formatted result breaks my code.
- [ ] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.
- [ ] I would like a release if this problem is solved.

#### Options

Fantomas main branch at 2024-09-14T12:20:28Z - 06d86ad91974467f0e82f5f0ca7d3c58eef7e859

Default Fantomas configuration

<sub>Did you know that you can ignore files when formatting by using a [.fantomasignore file](https://fsprojects.github.io/fantomas/docs/end-users/IgnoreFiles.html)?</sub>
<sub>PS: It's unlikely that someone else will solve your specific issue, as it's something that you have a personal stake in.</sub>

### PR DESCRIPTION
**Pull request by nojaf**
Fixes https://github.com/fsprojects/fantomas/issues/3131

- Moved some related tests around to new testfile.
- Targeting v7.0 because it can lead to a git diff when folks would update.

Could you review this one for me @brianrourkeboll? Just to make sure I didn't miss anything.
@dawedawe please review the Fantomas' side of things.

Thanks all!

### CODE REVIEW

File: src/Fantomas.Core.Tests/PrefixTests.fs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    [ "-86y"
+      "-86s"
+      "-86"
+      "-86l"
+      "-123n"
+      "-86L"
+      "-4.41F"
+      "-4.14"
+      "-12456I"
+      "-0.7833M" ]
```
Comment (brianrourkeboll) [inline src/Fantomas.Core.Tests/PrefixTests.fs:129]: It's probably worth adding a literal with a `+` here for clarity's sake, e.g., `+1`.

File: src/Fantomas.Core.Tests/PrefixTests.fs
Comment (brianrourkeboll) [inline src/Fantomas.Core.Tests/PrefixTests.fs:1]: We should probably add a test for quotations here as well?

E.g., the space must be kept in

```fsharp
let (~+) = id
let _ = + <@ 1 @>
```

Comment (brianrourkeboll) [reply src/Fantomas.Core.Tests/PrefixTests.fs:1]: Also for measures:

```fsharp
let _ = - +1<m>
let _ = - -1<m>
```

File: src/Fantomas.Core.Tests/PrefixTests.fs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        config
+    |> prepend newline
+    |> should
+        equal
+        """
+let _ = + <@ 1 @>
+"""
+
+[<Test>]
+let ``add space between prefix and measure`` () =
```
Comment (brianrourkeboll) [inline src/Fantomas.Core.Tests/PrefixTests.fs:198]:
```suggestion
let ``add space between prefix and measure literal that starts with a symbol`` () =
```

File: src/Fantomas.Core/CodePrinter.fs [hunk 3]
```diff
@@ -753,33 +753,40 @@ let genExpr (e: Expr) =
...
+            // E.g. !! @"foobar", because !!@ would be mistaken for an operator.
+            match constant with
+            | Constant.FromText(stn) ->
+                match stn.Text.[0] with
+                | '@'
+                | '-'
+                | '+' -> genWithSpace
+                | _ -> genWithoutSpace
+            // E.g. - +1<m>
+            | Constant.Measure _ -> genWithSpace
```
Comment (brianrourkeboll) [inline src/Fantomas.Core/CodePrinter.fs:772]: There doesn't need to be a space if the measure literal doesn't start with `+` or `-`, though.

File: src/Fantomas.Core.Tests/PrefixTests.fs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+let subtractTwo = + %s{literalValue}
+"""
+
+let operator_application_literal_values_without_sign =
+    [ "86uy"
+      "86us"
+      "86u"
+      "86ul"
+      "0x00002D3Fun"
+      "86UL"
```
Comment (brianrourkeboll) [inline src/Fantomas.Core.Tests/PrefixTests.fs:160]:
```suggestion
      "86UL"
      "1<m>
```

Or in a separate test if you like.

**Review by brianrourkeboll — APPROVED** [review]

**Review by dawedawe — APPROVED** [review]

**Pull request merged**